src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUseageDescriptorFactory.cs

```diff
@@ -0,0 +1,136 @@
+                {
+                    return null;
+                }
+            }
+
+            XmlDocumentationProvider documentationProvider = null;
+            if (!XmlDocumentationProviderCache.TryGetValue(typeAssemblyLocation, out documentationProvider))
+            {
+                var xmlDocumentationFile = GetXmlDocumentationFile(typeAssemblyLocation);
+
+                // We only want to process the file if it exists. In the case it doesn't, a null value will be added
+                // to the cache to not constantly look for new XML files.
+                if (xmlDocumentationFile != null)
+                {
+                    documentationProvider = new XmlDocumentationProvider(xmlDocumentationFile.FullName);
+                }
+
+                XmlDocumentationProviderCache.Add(typeAssemblyLocation, documentationProvider);
+            }
+
+            // Members will be null if there is no associated XML file for the provided typeAssembly.
+            if (documentationProvider != null)
+            {
+                if (documentationProvider.HasDocumentation(id))
+                {
+                    var summary = documentationProvider.GetSummary(id);
+                    var remarks = documentationProvider.GetRemarks(id);
+
+                    return new TagHelperUseageDescriptor(summary, remarks);
+                }
+            }
+
+            return null;
+        }
+
+        private static FileInfo GetXmlDocumentationFile(string typeAssemblyLocation)
+        {
+            if (string.IsNullOrWhiteSpace(typeAssemblyLocation) ||
+                typeAssemblyLocation.IndexOfAny(InvalidPathCharacters) != -1)
+            {
+                return null;
+            }
+
+            try
+            {
+                var assemblyDirectory = Path.GetDirectoryName(typeAssemblyLocation);
+                var assemblyName = Path.GetFileName(typeAssemblyLocation);
+                var assemblyXmlDocumentationName = Path.ChangeExtension(assemblyName, ".xml");
+                var culture = CultureInfo.CurrentCulture;
+
+                var assemblyXmlDocumentationFile = new FileInfo(
+                    Path.Combine(assemblyDirectory, assemblyXmlDocumentationName));
+
+                // If there's not an XML file side-by-side the .dll it may exist in a culture specific directory.
+                if (!assemblyXmlDocumentationFile.Exists)
+                {
+                    var fallbackDirectories = GetCultureFallbackDirectories();
+                    assemblyXmlDocumentationFile = fallbackDirectories
+                        .Select(fallbackDiretory =>
+                            new FileInfo(
+                                Path.Combine(assemblyDirectory, fallbackDiretory, assemblyXmlDocumentationName)))
+                        .FirstOrDefault(file => file.Exists);
+                }
+
+                return assemblyXmlDocumentationFile;
+            }
+            catch (PathTooLongException)
```

Comment:
Instead of doing this could a custom exception thrown by the `XmlDocumentationProvider` that can be caught OR could add a property to the provider like `XmlFileResolved`. Problem with adding a property would then add the requirement to throw whenever one of the `GetSummary` or `GetRemarks` methods is invoked.

Adding this comment for discussion.